scapy/contrib/lldp.py

```diff
@@ -698,6 +711,597 @@ class LLDPDUGenericOrganisationSpecific(LLDPDU):
+    POWER_TYPE_NO = {
+        1: 'type 1',
+        0: 'type 2',
+    }
+
+    # IEEE802.3bt-2018 - sec. 79.3.2.4
+    POWER_TYPE_DIR = {
+        1: 'PD',
+        0: 'PSE',
+    }
+
+    # IEEE802.3bt-2018 - sec. 79.3.2.4
+    POWER_SOURCE_PD = {
+        0b11: 'PSE and local',
+        0b10: 'reserved',
+        0b01: 'PSE',
+        0b00: 'unknown',
+    }
+
+    # IEEE802.3bt-2018 - sec. 79.3.2.4
+    POWER_SOURCE_PSE = {
+        0b11: 'reserved',
+        0b10: 'backup source',
+        0b01: 'primary source',
+        0b00: 'unknown',
+    }
+
+    # IEEE802.3bt-2018 - sec. 79.3.2.4
+    PD_4PID_SUP = {
+        0: 'not supported',
+        1: 'supported',
+    }
+
+    # IEEE802.3bt-2018 - sec. 79.3.2.4
+    POWER_PRIO = {
+        0b11: 'low',
+        0b10: 'high',
+        0b01: 'critical',
+        0b00: 'unknown',
+    }
+
+    fields_desc = [
+        BitEnumField('_type', 127, 7, LLDPDU.TYPES),
+        BitField('_length', 12, 9),
+        ThreeBytesField('org_code', LLDPDUGenericOrganisationSpecific.ORG_UNIQUE_CODE_IEEE_802_3),  # noqa: E501
+        ByteField('subtype', 2),
+        FlagsField('MDI_power_support', 0, 8, LLDPDUPowerViaMDI.MDI_POWER_SUPPORT),
+        ByteEnumField('PSE_power_pair', 1, LLDPDUPowerViaMDI.PSE_POWER_PAIR),
+        ByteEnumField('power_class', 1, LLDPDUPowerViaMDI.POWER_CLASS),
+        BitEnumField('power_type_no', 1, 1, POWER_TYPE_NO),
+        BitEnumField('power_type_dir', 1, 1, POWER_TYPE_DIR),
+        MultipleTypeField([
+            (
+                BitEnumField('power_source', 0b01, 2, POWER_SOURCE_PD),
+                lambda pkt: pkt.power_type_dir == 1
+            ),
+        ], BitEnumField('power_source', 0b01, 2, POWER_SOURCE_PSE)),
+        MultipleTypeField([
+            (
+                BitEnumField('PD_4PID', 0, 2, PD_4PID_SUP),
+                lambda pkt: pkt.power_type_dir == 1
+            ),
+        ], BitField('PD_4PID', 0, 2)),
+        BitEnumField('power_prio', 0, 2, POWER_PRIO),
+        ShortField('PD_requested_power', None),
+        ShortField('PSE_allocated_power', None),
+    ]
+
+    def do_build(self):
+        # Power is in tenths of Watts
+        # IEEE802.3bt-2018 - sec. 79.3.2.{5,6}
+        conv_fields = ['PD_requested_power', 'PSE_allocated_power']
+        backup_fields = [(field, getattr(self, field)) for field in conv_fields]
+        for field in conv_fields:
+            val = int((getattr(self, field) or 0) * 10)
+            setattr(self, field, val)
+        s = super(LLDPDUPowerViaMDIDDL, self).do_build()
+        for field, val in backup_fields:
+            setattr(self, field, val)
+        return s
+
+    def post_dissect(self, s):
+        # Power is in tenths of Watts
+        # IEEE802.3bt-2018 - sec. 79.3.2.{5,6}
+        conv_fields = ['PD_requested_power', 'PSE_allocated_power']
+        s = super(LLDPDUPowerViaMDIDDL, self).post_dissect(s)
+        for field in conv_fields:
+            val = getattr(self, field) / 10
```

Comment:
This can be removed if you use a ScalingField

Reply:
Fixed